helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -68,23 +71,47 @@ public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebala
     List<AssignableNode> nodes = new ArrayList<>(clusterModel.getAssignableNodes().values());
     Set<String> busyInstances =
         getBusyInstances(clusterModel.getContext().getBestPossibleAssignment().values());
-    // Sort the replicas so the input is stable for the greedy algorithm.
-    // For the other algorithm implementation, this sorting could be unnecessary.
-    for (AssignableReplica replica : getOrderedAssignableReplica(clusterModel)) {
+    List<AssignableReplica> toBeAssignedReplicas = clusterModel.getAssignableReplicaMap().values().stream()
+        .flatMap(replicas -> replicas.stream()).collect(Collectors.toList());
+    // Compute overall utilization of the cluster. Capacity dimension -> <total usage, total capacity>
+    Map<String, MutablePair<Integer, Integer>> overallClusterUtilMap =
```

Comment:
The Pair usage makes it very unclear what each number represents, especially when this map is passed into later functions. The references of `left` and `right` has no obvious meaning later on. 
I suggest:
1. getting rid of the `total capacity`. As I commented later, seems like the calculation cancels out `total capacity` and it may not be needed at all;
2. creating helper functions in ClusterModel with meaningful names, and leave all calculations/updates in the helper functions;
3. using 2 maps with meaningful names.